tests/integration/high_availability/test_upgrade.py

```diff
@@ -0,0 +1,101 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+import logging
+
+import pytest
+from integration.helpers import get_primary_unit, retrieve_database_variable_value
+from integration.high_availability.high_availability_helpers import (
+    ensure_all_units_continuous_writes_incrementing,
+    high_availability_test_setup,
+)
+from lightkube import Client
+from lightkube.resources.apps_v1 import StatefulSet
+from pytest_operator.plugin import OpsTest
+
+logger = logging.getLogger(__name__)
+
+TIMEOUT = 15 * 60
+
+
+@pytest.mark.group(1)
+async def test_build_and_deploy(ops_test: OpsTest) -> None:
+    """Simple test to ensure that the mysql and application charms get deployed."""
+    await high_availability_test_setup(ops_test)
+
+
+@pytest.mark.group(1)
+@pytest.mark.abort_on_fail
+async def test_pre_upgrade_check(ops_test: OpsTest) -> None:
+    """Test that the pre-upgrade-check action runs successfully."""
+    mysql_app_name, _ = await high_availability_test_setup(ops_test)
+
+    mysql_units = ops_test.model.applications[mysql_app_name].units
+
+    logger.info("Get leader unit")
+    leader_unit = None
+    for unit in mysql_units:
+        if await unit.is_leader_from_status():
+            leader_unit = unit
+            break
+
+    assert leader_unit is not None, "No leader unit found"
+    logger.info("Run pre-upgrade-check action")
+    action = await leader_unit.run_action("pre-upgrade-check")
+    await action.wait()
+
+    logger.info("Assert slow shutdown is enabled")
+    for unit in mysql_units:
+        value = await retrieve_database_variable_value(ops_test, unit, "innodb_fast_shutdown")
+        assert value == 0, f"innodb_fast_shutdown not 0 at {unit.name}"
+
+    primary_unit = await get_primary_unit(ops_test, leader_unit, mysql_app_name)
+
+    assert primary_unit.name == f"{mysql_app_name}/0", "Primary unit not set to unit 0"
+
+    logger.info("Assert partition is set to 2")
+    client = Client()
+    statefulset = client.get(
+        res=StatefulSet, namespace=ops_test.model.info.name, name=mysql_app_name
+    )
+
+    assert statefulset.spec.updateStrategy.rollingUpdate.partition == 2, "Partition not set to 2"
+
+
+@pytest.mark.group(1)
+@pytest.mark.abort_on_fail
+async def test_upgrade_charms(ops_test: OpsTest, continuous_writes) -> None:
+    mysql_app_name, _ = await high_availability_test_setup(ops_test)
+    logger.info("Ensure continuous_writes")
+    await ensure_all_units_continuous_writes_incrementing(ops_test)
+
+    # 8.0.33
+    resources = {
+        "mysql-image": "ghcr.io/canonical/charmed-mysql@sha256:753477ce39712221f008955b746fcf01a215785a215fe3de56f525380d14ad97"
+    }
```

Comment:
nit: we should really modify `high_availability_test_setup` to take an optional kwarg that will result in deployment of < v8.0.33, and then we can always use the hash in metadata to ensure that there will be an upgrade to >= v8.0.33. this can be a follow up if you'd like